daemon/src/db/failed.rs

```diff
@@ -0,0 +1,179 @@
+//! This module allows us to move faile CFDs to a separate table so
+//! that we can reason about loading them independently from open
+//! CFDs.
+//!
+//! Therefore, it also provides an interface to load failed CFDs: the
+//! `FailedCfdAggregate` trait. Implementers of the trait will be able
+//! to call the `crate::db::load_all_cfds` API, which loads all types
+//! of CFD.
+
+use crate::db;
+use crate::db::delete_from_cfds_table;
+use crate::db::delete_from_events_table;
+use crate::db::load_cfd_row;
+use crate::db::CfdAggregate;
+use crate::db::Connection;
+use anyhow::Result;
+use model::Contracts;
+use model::Identity;
+use model::Leverage;
+use model::OrderId;
+use model::Position;
+use model::Price;
+use model::Role;
+use sqlx::Connection as _;
+use sqlx::Sqlite;
+use sqlx::Transaction;
+
+/// A trait for building an aggregate based on a `FailedCfd`.
+pub trait FailedCfdAggregate: CfdAggregate {
+    fn new_failed(args: Self::CtorArgs, cfd: FailedCfd) -> Self;
+}
+
+/// Data loaded from the database about a failed CFD.
+#[derive(Debug, Clone, Copy)]
+pub struct FailedCfd {
```

Comment:
🛠 Failed CFDs may have events, i.e. we may have a `ContractSetupStarted` and `ContractSetupFailed` event. I think it might be a good idea to save the timestamp of the first event recorded - that is what we use for sorting the open CFDs.. 
I think a user would still expect to see the CFDs chronologically ordered, but without timestamps we cannot really achieve that. 
I think this is actually already missing when processing closed cfds - I would always save it as `creation_timestamp` (or `first_event_timestamp` it you want). I am not sure we have to add it as another field, we could also load it from the eventlog (I suppose failed will also have eventlog).
I think we should always have at least one event in the database, so we should always be able to come up with a timestamp. For cases where we don't have any events we could actually just delete from `cfds` because this is too incomplete to do anything.

Reply:
> 🛠 Failed CFDs may have events, i.e. we may have a ContractSetupStarted and ContractSetupFailed event.

Do you think it's worth saving events for failed CFDs? To me it's even questionable if we should be persisting them! One could imagine that a failed CFD could eventually just become an ephemeral notification for the user.

But with this PR I would persist them since we have no better alternative at the moment.

> I think it might be a good idea to save the timestamp of the first event recorded - that is what we use for sorting the open CFDs..
I think a user would still expect to see the CFDs chronologically ordered, but without timestamps we cannot really achieve that.

Good point: we want these to appear in order. It is a bit annoying to have to process the events at all just to get a timestamp, but I guess we should.

> I think this is actually already missing when processing closed cfds - I would always save it as creation_timestamp (or first_event_timestamp it you want).

It's true that the closed CFDs are not sorted properly (because we don't use the `projection::Cfd::aggregate` field. More importantly, the `profit_percent`, `profit_btc` and `payout` fields are actually wrong. I added a test to verify this, but I did not realise that we were calling `with_current_quote` on all CFDs (doing so on closed CFDs overwrites these values with wrong ones).

> I am not sure we have to add it as another field, we could also load it from the eventlog (I suppose failed will also have eventlog).

I would prefer not to add entries to the `event_log` table for failed or rejected CFDs, because they're always going to be the same. I think this information wouldn't be very useful.

Reply:
> I would prefer not to add entries to the event_log table for failed or rejected CFDs, because they're always going to be the same. I think this information wouldn't be very useful.

I am not sure about that. In *most* cases it won't have value, but in some cases it might. There might be error scenarios where something goes into a weird failed state that should be analysed. It's not expensive to save these event log entries. 
If we decide to not keep an eventlog for failed at all we *definitely* need a timestamp, because otherwise we don't know anything about the time the CFD "failed".